app/ts-cli/geminicli/export.go

```diff
@@ -0,0 +1,1001 @@
+	var offset int64 = 0
+	var rows []influx.Row
+	for {
+		rows, offset, recordCompBuff, err = e.readWalRows(fd, offset, fileSize, recordCompBuff)
+		if err != nil {
+			if err == io.EOF {
+				return nil
+			}
+			return nil
+		}
+		if err = e.writeRows(rows, outputWriter); err != nil {
+			return err
+		}
+	}
+}
+
+// readWalRows read some rows from the fd, and reuse recordCompBuff to save memory.
+func (e *Exporter) readWalRows(fd fileops.File, offset, fileSize int64, recordCompBuff []byte) ([]influx.Row, int64, []byte, error) {
+	if offset >= fileSize {
+		return nil, offset, recordCompBuff, io.EOF
+	}
+
+	// read record header
+	var recordHeader [engine.WalRecordHeadSize]byte
+	n, err := fd.ReadAt(recordHeader[:], offset)
+	if err != nil {
+		e.stderrLogger.Println(errno.NewError(errno.ReadWalFileFailed, fd.Name(), offset, "record header").Error())
+		return nil, offset, recordCompBuff, io.EOF
+	}
+	if n != engine.WalRecordHeadSize {
+		e.stderrLogger.Println(errno.NewError(errno.WalRecordHeaderCorrupted, fd.Name(), offset).Error())
+		return nil, offset, recordCompBuff, io.EOF
+	}
+	offset += int64(len(recordHeader))
+
+	// prepare record memory
+	compBinaryLen := binary.BigEndian.Uint32(recordHeader[1:engine.WalRecordHeadSize])
+	recordCompBuff = bufferpool.Resize(recordCompBuff, int(compBinaryLen))
+
+	// read record body
+	var recordBuff []byte
+	n, err = fd.ReadAt(recordCompBuff, offset)
+	if err == nil || err == io.EOF {
+		offset += int64(n)
+		var innerErr error
+		recordBuff, innerErr = snappy.Decode(recordBuff, recordCompBuff)
+		if innerErr != nil {
+			e.stderrLogger.Println(errno.NewError(errno.DecompressWalRecordFailed, fd.Name(), offset, innerErr.Error()).Error())
+			return nil, offset, recordCompBuff, io.EOF
+		}
+		var rows []influx.Row
+		var tagPools []influx.Tag
+		var fieldPools []influx.Field
+		var indexKeyPools []byte
+		var indexOptionPools []influx.IndexOption
+		var err error
+
+		rows, _, _, _, _, innerErr = influx.FastUnmarshalMultiRows(recordBuff, rows, tagPools, fieldPools, indexOptionPools, indexKeyPools)
+
+		if innerErr == nil {
+			return rows, offset, recordCompBuff, err
+		}
+		return rows, offset, recordCompBuff, innerErr
+	}
+	e.stderrLogger.Println(errno.NewError(errno.ReadWalFileFailed, fd.Name(), offset, "record body").Error())
+	return nil, offset, recordCompBuff, io.EOF
+}
+
+// writeRows process a cluster of rows
+func (e *Exporter) writeRows(rows []influx.Row, outputWriter io.Writer) error {
+	for _, r := range rows {
+		if err := e.writeSingleRow(r, outputWriter); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// writeSingleRow parse a single row to lint protocol, and writes it.
+func (e *Exporter) writeSingleRow(row influx.Row, outputWriter io.Writer) error {
+	measurementWithVersion := row.Name
+	var measurementName string
+	measurementName = influx.GetOriginMstName(measurementWithVersion)
+	measurementName = EscapeMstName(measurementName)
```

Comment:
Consider the reuse of frequent memory applications.

Reply:
In the leaste commit, I tried it on `writeSingleRecord()` and  `writeSingleRow()`